core/iwasm/common/gc/gc_common.c

```diff
@@ -165,31 +199,68 @@ wasm_struct_obj_new_with_type(WASMExecEnv *exec_env, WASMStructType *type)
 }
 
 WASMArrayObjectRef
-wasm_array_obj_new_with_typeidx(WASMExecEnv *exec_env, uint32 type_idx)
+wasm_array_obj_new_with_typeidx(WASMExecEnv *exec_env, uint32 type_idx,
+                                uint32 length, wasm_value_t *init_value)
 {
-    /* TODO */
-    return NULL;
+    WASMModuleCommon *module = wasm_exec_env_get_module(exec_env);
+    WASMType *defined_type = wasm_get_defined_type(module, type_idx);
+    WASMRttTypeRef rtt_type = wasm_runtime_malloc(sizeof(WASMRttType));
+    rtt_type->type_flag = defined_type->type_flag;
+    rtt_type->inherit_depth = defined_type->inherit_depth;
+    rtt_type->defined_type = defined_type;
+    rtt_type->root_type = defined_type->root_type;
+
+    WASMArrayObjectRef array_obj =
+        wasm_array_obj_new(exec_env, rtt_type, length, init_value);
+    return array_obj;
 }
 
 WASMArrayObjectRef
-wasm_array_obj_new_with_type(WASMExecEnv *exec_env, WASMArrayType *type)
+wasm_array_obj_new_with_type(WASMExecEnv *exec_env, WASMArrayType *type,
+                             uint32 length, wasm_value_t *init_value)
 {
-    /* TODO */
-    return NULL;
+    WASMRttTypeRef rtt_type = wasm_runtime_malloc(sizeof(WASMRttType));
```

Comment:
Should use `wasm_rtt_type_new` API rather than directly call wasm_runtime_malloc

Reply:
The API `wasm_rtt_type_new` requires a parameter `defined_type_idx`, in this funtion, we only know `type` not its index. And it seems that we can't get `type_index` based on its `type`.

Maybe we should add a new API `wasm_rtt_type_new_with_type`?

Reply:
One way is to get `WASMModule *module` from exec_env, then traverse `module->types[i]` to check whether `type == module->types[i]` to get it. Another way is to store `type_idx` as a common field in WASMType/WASMStructType/WASMFuncType/WASMArrayType, set it in the wasm loader.

Reply:
Get~
Since the normalization has been done, we can use method 1 here, and method 2 requires changes in other files, maybe we need to consider the necessity of store `type_idx` as a common field, and create another PR to do it.